### DISCUSSION TITLE
Some questions for further improvement

### DISCUSSION THREAD

**Discussion by SvenRuin** [grafana/grafana-csv-datasource#28] [Q&A] [answered]
Thank you for the great work you have done Marcus, starting the development of CSV plugin. Here are some questions from the perspective of a new user of Grafana:
1. Would it be possible that the names of the fields (or labels) are identified automatically based on the second line in the csv file? (For some csv files, that is a standard format, and in that case the units are on the third line.)
2. Would it be possible to speed up the handling of large csv files, where it can be of interest to show many data columns (fields) in the same graph?
3. Would it be possible to avoid automatic updateing of graphs (automatic updating seems to have crashed Linux for me once when I was editing a panel)?
4. When I try to zoom in on a certain time period in a graph, it seems to go to the wrong time - has anyone else experienced that?
5. When I in a dashboard select e.g. "Last 24 hours", the last 24 hours are not shown - has anyone else experienced that?

**Comment by marcusolsson** [comment, marked as answer]
Thank you for the kind words!

1. Setting units automatically would be really useful. I know there's been discussions for data source plugins to return "hints" for how to display the data. For example, what unit to use. Unfortunately, I don't believe that has been implemented yet. 
2. Things that I could do to improve the performance is to look for opportunties when downloading, and processing of data server-side. However, it sounds like the poor performance you're seeing seems to be dependent on the visualization. We could add server-side caching to avoid having to download the file, but I can't guarantee that will make it easier to work with.
3. I think we should look into reducing the data being processed by the visualization. For starters, have you tried enabling "Ignore unknown" and only adding the columns you're interested in to the schema?
4. The plugin currently doesn't filter data based on the time interval, since there could be multiple columns with timestamps, in which case you'd have to pick the column to filter on. A couple of improvements come to mind:
   - If there's only one time field, use that for filtering
   - If there are multiple time fields, add a dropdown to select the primary time field

Let me know if any of the suggestions would be of help to you!

**Reply by SvenRuin** [reply]
Now I have looked more at this. Regarding my points numbered 2-3, I believe the problems I experiened occured because I was using a large csv file when setting up or editing a panel. Thus, I believe I can in these cases simply use a shorter csv file and then switch to the larger csv file when the panel is set up right. Regarding my points numbered 4-5, these problems don't seem to occur for me any longer. Don't know why.

**Comment by marcusolsson** [comment]
I've done some initial performance tests on the server-side processing with a 30MB CSV file with ~300k lines, which takes less than a second to complete, so there's not much to do there. 

Even when loading the same file locally, the UI freezes. However, when I check Ignore unknown and only select a subset of the columns by adding them to the schema, performance improves. 

I think the solution is to reduce the data that's being visualized:

- We can improve the user experience by capping the number of lines that is being sent to the browser and show a warning that data has been truncated.
- We can introduce some form of query language. We could look into using https://github.com/mithrandie/csvq to support SQL-like queries to let users filter the data they want to visualize.

**Reply by SvenRuin** [reply]
At this time, I think we don't need to worry any more about my points numbered 2-5.

**Comment by SvenRuin** [comment]
Two more points:
6. Would it in the data source settings be possible to allow multiple files to be read (and treated as one data source)? An example: I would like to define the data source "Data2018-202X" to first consist of "Data2018.dat", "Data2019.dat" and "Data2020.dat". Later, "Data2021.dat" (and so on) would be added to the list of files.
7. If so, would it be possible that new fields (columns) are added in later files, without causing errors? An example: "Data2018.dat" consists of two fields (say columns "Date" and "Total Sales"). "Data2019.dat" consists of four fields (the same columns as before, plus "Sales Region 1" and "Sales Region 2"). "Data2020.dat" consists of five fields (new column "Sales Region 3"). Perhaps "not a number" could be used for the missing data in the first files, so if a graph is made of "Total Sales" and "Sales Region 3", then "Total Sales" would be shown for the entire time period but the line for "Sales Region 3" would start in 2020.
